### DISCUSSION TITLE
GoB_v3_5_56

### DISCUSSION THREAD

**Discussion by kromar** [JoseConseco/GoB#258] [General]
- added manual import option

<hr /><em>This discussion was created from the release <a href='https://github.com/JoseConseco/GoB/releases/tag/v3_5_56'>GoB_v3_5_56</a>.</em>

**Comment by KptChaos** [comment]
I'm looking for information how to work with that addon. I can't find information what kind of selection groups should I create within Blender to be have them converted into polygroups within Zbrush...

**Comment by kromar** [comment]
you can use vertex groups or face maps to send over to zbrush as polygroups at the moment. Im working on adding material to polygroups support but that will take a bit until its ready.

**Comment by sidney-eliot** [comment]
Why is the slider limited to 0.1 - 10. There is no reason to have a cap, please fix (Manually typing number also doesn't work)

Edit: And yes I know that I know that GOB uses zbrushes export scale, but maximum user flexibility is still better

Edit 2: Changing export scale in ZBrush doesn't work because when importing it back to ZBrush the scale is not the same anymore. This means that one has to use the GOB addon scale slider which if one uses the optimal ZBrush scale of 2 units only allows for object of the size 0.02m - 20m to be exported to blender. For bigger creatures this is not enough

**Reply by kromar** [reply]
> Why is the slider limited to 0.1 - 10. There is no reason to have a cap, please fix (Manually typing number also doesn't work)
> 
> Edit: And yes I know that I know that GOB uses zbrushes export scale, but maximum user flexibility is still better
> 
> Edit 2: Changing export scale in ZBrush doesn't work because when importing it back to ZBrush the scale is not the same anymore. This means that one has to use the GOB addon scale slider which if one uses the optimal ZBrush scale of 2 units only allows for object of the size 0.02m - 20m to be exported to blender. For bigger creatures this is not enough

All the sliders should allow to manually type a bigger value, i just tested the manual and zunit export scale since i assume that is what you are talking about and i can enter 100 manually to override the soft max of 10. Is it a other value that has a hard limit?

There are some things in Zbrush which i can not control with this addon without creating a dedicated GoB exporter in zbrush. Im aware of some scaling issues and am trying to overcome some of those but its not a simple thing to do.
Can you post the exact steps you are performing which result in units not work as expected for you so i can see if anything could be done?

**Reply by sidney-eliot** [reply]
The slider I am talking about is the slider in: Blender>GOB addon settings>Options>Scale Factor (with Scale set to manual). When typing a number higher or lower than the max/ min it jumps back to the highest or lowest value. The version of GOB I am using is 3.5.67. 

---

As to the scale not exporting you probably can't change it because you can't change the GOZ stuff without releasing your own GOZ. 


As mentioned there are 2 ways of controlling scale, using the scale slider in the blender addon of yours or in the scale export tab of ZBrush. The reason why using the export scale slider in Zbrush doesn't work is, when one exports the model from ZBrush to blender with a export scale of let's say to 2 and a scale of 1 in your GOB addon. Then it will import into blender nicely. When one then exports it back to blender using the GOB export button, then the scale will be off because ZBrush doesn't know anymore that it had scaled that object by to at export. So every time one export between ZBrush and blender, the model will get bigger and bigger and bigger. This problem does not occur when using ZBrush scale export at 1 and changing the blender GOB addon scale slider. That's why it's so important this slider is not caped.


Oh and thanks so much for this addon, I once had the opinion: Either I have ZBrushes tools or Blenders tools. Now I can fluently use them all.

**Reply by kromar** [reply]
> The slider I am talking about is the slider in: Blender>GOB addon settings>Options>Scale Factor (with Scale set to manual). When typing a number higher or lower than the max/ min it jumps back to the highest or lowest value. The version of GOB I am using is 3.5.67.

i just released a new version where i fixed quite a bit of ui caps, 3_5_85 https://github.com/JoseConseco/GoB/releases/tag/v3_5_85 maybe the cap was still there in a older version so would be great if you could give it a try and see if it still resets to 10.

There is also a fix i made for some scaling but i do not know if it also relates to the issue you describe with scaling, maybe you can see if it changes the behavior for you.

I will have to try to reproduce that scaling issue outside of scale 1 for manual scaling you describe to see if its possible to fix it. Im using mostly the blender units scaling so i dont come across these other issues. Thanks for bringing it up, the addon can only be as good as the feedback that comes from the users to improve on it!

**Reply by sidney-eliot** [reply]
Will do.

Do you know why ZBrush goes through every subtool when I click the GOZ button? I would think it would only have to look at the selected subtool

**Reply by kromar** [reply]
I dont know why they do it, maybe some kimd of limitation how to find the selected subtool. I had a similar issue on import which i had to workaround.

**Reply by sidney-eliot** [reply]
Well the upper limit with 3.5.85 is now un caped. However for some reason the lower limit still has the issue from before. The smallest value is 0.1. And this isn't a blender problem because the custom properties slider which is the same one as you are using can go down quite far.

Something that has been bugging me has now also been fixed, that the 3 GOB buttons are now at the top right which makes them not disappear when pulling the blender windows smaller👍

I haven't tried the scale thing yet, I'll do that when I have some time

**Comment by sidney-eliot** [comment]
And is there a reason why ZBrush export offset doesn't work with GOB

**Reply by kromar** [reply]
> 

Can you explain what Zbrush export offset is?

**Reply by sidney-eliot** [reply]
ZBrush has a export scale multiplier as you know and below that slider there are 3 axis offset sliders, one for each axis. Let's say we export a character, with all 3 offsets set to 0. The character will export to blender at some random position on the up axis. But when one sets the ZBrush Y axis(up) offset to 1 then the character exports to blender with the feet nicely resting at the world origin. For this to work the floor elevation in ZBrush has to be set to -1 so it dynamically adjusts to the lowest subtool in the tool. 

Now you might think, why doesn't one simply place the characters lowest point inside of ZBrush at the 0 0 0 point instead of using dynamic floor, the reason for that is: Let's say we want to extend the legs of the character, if we do that we would have to move up the entire character again which would be very tedious, that's why dynamic floor is so good. And going back to export offset, when the value is set to 1 then it nicely plays together with the -1  floor elevation to set the characters feet on the floor when exported to blender

**Reply by kromar** [reply]
oh that is a neat trick, i was not aware that this is how the dynamic floor in zbrush works and could be used. I will see what i can do in that regard, sounds like a very useful functionality to have.

**Reply by kromar** [reply]
i created 2 issues to look into what you reported, i dont know when i get around to look into it but you can follow it over here:
https://github.com/JoseConseco/GoB/issues/336
https://github.com/JoseConseco/GoB/issues/337

**Reply by sidney-eliot** [reply]
Be warned it is very complex. Things like what if I export a model to blender that is finished, I retopo the model and so on, and then I decide to add or change a piece of the mesh. Then we open ZBrush and let's say, make the leg a bit longer. Then the floor will change which could be a massive issue because now the model doesn't line up anymore. Eyeballing or snapping tools can fix this though, but still. Things like this need to be considered when adding this feature.

**Reply by kromar** [reply]
Im fully aware of the complexity of the issue, most likely some thing will need to be left to the user to do the decision. I will let you know when i have a better understanding of it what is possible so a good solution can be found.

**Comment by sidney-eliot** [comment]
Maybe consider changing some of the default values in the addon. Things like weld at 0.0001 in my opinion should be off per default. It is quite common to have perfectly overlapping verts when working with creases controlled by a subD or bevel like this:
![image](https://user-images.githubusercontent.com/85735034/178158786-c6c78710-1b18-4f0f-b757-1a6c88c7aaf8.png)

If one doesn't look at all the settings this can lead to a very frustrating time.

---

It would also be cool if you could make a GOB official doc with things like best practice workflows. I could definitely help with that. Everything publicly available about this addon and general workflows between blender and zbrush is really bad not even that video from flipped normals is that good

**Reply by kromar** [reply]
![image](https://user-images.githubusercontent.com/1472884/178159554-44595191-7f9f-49a7-9d42-cc2f9ebd17d3.png)
would you also disable the deletion of non manifolds by default? (this was implemented to avoid issues with disconnected vertices)

I started to write a wiki some time ago but its very tedious to do this on github and its lacking most of the features. It would be great to get some assistance on the documentation since i barely have enough time to work on the code. Maybe a different documentation page would be better

**Reply by sidney-eliot** [reply]
Yes I do think that should be disabled but I wasn't sure if that is good to be off per default. Could you elaborate on what it does. And I think renaming it could be good because if it actually deletes non manifold faces it would be destroying the entire mesh on open faced models. 

Example:
![image](https://user-images.githubusercontent.com/85735034/178160449-fdee122f-b808-44d7-8363-b91cc1fabece.png)
Blender says that this mesh has 32 non manifold edges even though it's a perfectly fine mesh. You can see what blender considers as non manifold edges with the 3D print addon, which checks mesh integrity.
The result of this mesh was: 
![image](https://user-images.githubusercontent.com/85735034/178160452-be9c9b4e-4c8d-43bb-bc50-811580512600.png)




---

Because of documentation I do have quite a bit of experience with that and find that git hub's wikis are quite fine for that. Yes some things it lacks but all in all it's good, not to say that it's the best

**Reply by kromar** [reply]
you are right the name should change since its only looking for disconnected vertices/ edges without faces which zbrush can not import
this would be considered a loose element to be removed but the edge 
![image](https://user-images.githubusercontent.com/1472884/178160942-8b550291-42df-4e8d-a9cb-8a8e8d9c8aa9.png)

now that im looking at this it seems this is not working as intended any more, maybe its not required any longer and can be removed

EDIT:

after looking up the ticket i confused this with a import issue https://github.com/JoseConseco/GoB/issues/210 https://github.com/JoseConseco/GoB/issues/104

for the exporting non manifold, it removes internal non manifold faces. If you have a proposal how to name it let me know.
![image](https://user-images.githubusercontent.com/1472884/178161342-82895110-edd5-485b-8166-42a8df753774.png)

**Reply by sidney-eliot** [reply]
Well I can't say to much to these auto export fixes. It's stuff I would never use because it takes away control from me and makes issues because stuff is done in the background without me knowing it. And when one follows clean modeling conventions one will anyway never run into such issues. 

But if it should have export fixes I would rather do it as a message like: Your mesh has manifold faces, would you like to fix it (fix and cancel button below). This would be great for beginners and pros, beginners would click yes, have their mesh fixed and also know their model was tampered with and pros would click away the message and quickly fix the issue. Or if you implement the fixes nicely also use the fix button.

The downside of this you might conclude is that: It makes the export process take longer. Would say it's the opposite, because when GOB tampers with the mesh without asking and that then results in issues, then that steals much more time than clicking one button at export when the mesh has issues.

This actually happened to me before. I forgot that it auto merged and had to spend 10 min ripping the merged verts of my mesh apart

I'll give you a list of things that could have fix messages/ settings toggles and the correct names

**Comment by sidney-eliot** [comment]
# ESSENTIAL
These things are a must for the model to properly export to ZBrush

## Ngons
ZBrush can't work with faces that aren't either a tris or quad



# GOOD TO HAVE
These things won't break the export but should never be in the model, so one can might as well help the user by informing them about this and maybe even fixing it.

## Loose Geometry
There is basically no reason ever to have loose vertices in a model when one is about to sculpt on it. Blender only considers(select menu) 1 vertex as loose geometry so you have to decide if this should remove only single verts or connected ones as well
![image](https://user-images.githubusercontent.com/85735034/178207036-ee91dca6-a61c-480e-9011-9a9ddbafea76.png)

And loose faces are totally fine and common
![image](https://user-images.githubusercontent.com/85735034/178207798-175e6f13-282d-4427-9c68-eb891154f318.png)

## Interior Faces
This will make sculpting and other tools a nightmare. Vanilla blender can find these Select>Select by trait>Interior faces.
![image](https://user-images.githubusercontent.com/85735034/178208353-41d6dc86-1bff-402d-b423-5597cbc75eb3.png)

Note that this doesn't just have to be faces totally enclosed in the model. Things like this(below) would also not be desired, though there could be rare situation where one could actually use something like this in game ready assets.

![image](https://user-images.githubusercontent.com/85735034/178208735-4631b902-6d8e-4150-aee3-3cfc5de0c794.png)

An example of where this could be fine is the image below(Don't quote me on this I'm not sure maybe it is best to always avoid this). One could simply split the vertices so the face inside is not connected to the rest, making it not an interior face anymore. And leaving it connected could maybe also result in baking issues, so probably this is never desired.

**Edit:** Because this topology is not meant to be game ready but sculpting friendly, this definitely is a no no
![image](https://user-images.githubusercontent.com/85735034/178209233-c31b8605-2e83-483c-86e7-a609346bf8ab.png)


What this means is that you can't immediately assume that faces with flipped normals beside each other is bad.
![image](https://user-images.githubusercontent.com/85735034/178209655-66359489-eeb7-4690-99f9-f1d2f0b25d04.png)



## Bad Contiguous Edges
This plays right into that from above, faces beside each other with different normal directions. If you do implement this maybe give it a more user friendly name, the header name is the correct name but somethings like Bad Contiguous Normals or somethings else might be better.

## Intersecting faces
![image](https://user-images.githubusercontent.com/85735034/178210716-34fa2dc8-ba4a-44ff-b3f2-606fa81abd85.png)

## Overlapping geometry
Definitely worth warning but can often be wanted.

## Zero Edges/ Faces
This happens when an extrusion isn't pulled out, making edges and faces that have 0 volume
![image](https://user-images.githubusercontent.com/85735034/178211765-b5ab4345-086b-42ef-b57c-90910182b27c.png)



# THINGS TO MAYBE CONSIDER ADDING

### Name me
![image](https://user-images.githubusercontent.com/85735034/178229825-7926eaba-427d-40a1-afe1-dbb9c9ce65e8.png)

### Non Flat faces
This is an iffy one, I would rather not implement it but maybe you think it's good. The reason why it's iffy is where would one set the value of un flatness to be considered bad.
An extreme example of un flat faces:
![image](https://user-images.githubusercontent.com/85735034/178215131-2130bdd7-7bd5-4c58-8798-bc179325c8c5.png)



# Conclusion
I'll edit this comment when I come up with more.

So you can easily implement these checks, look into the 3d print addon, allot of the code for the checks is already done there.

Another thing is that all these messages should be toggleable in the addon pref tab, I for example would disable the overlapping verts message but it should be on per default (Edit: This is probably not needed when you go with a export window that shows it all at once)

# The export window
Why auto fixing is not good: With auto fix one has to decide to toggle it in the pref or not with a window like this one is always in the loop with the issues but can chose to ignore them.


I noticed that having many messages wouldn't be very user friendly so here is a design that shows all issues in one window
![image](https://user-images.githubusercontent.com/85735034/178218769-ec22e136-9ded-4df7-81a9-8d8cd1592a86.png)

The no fix button is for issues that can only be fixed by hand. For these issues if possible one can the let the user select the geo so they immediately know where the issue is.

# Edit 1: 
Intersecting faces might not be a good idea to add, zero edge faces might also be an iffy one

**Comment by sidney-eliot** [comment]
Ok so I looked a bit more into documentations and I think this is the best choice https://readthedocs.org/. Most addons use it. And they host it, quote from them: `We will host your documentation for free, forever. There are no tricks. We help over 100,000 open source projects share their docs, including a custom domain and theme.`

**Comment by kromar** [comment]
Thanks a lot for the detailed explanation and images to go with it! I agree that a popup with the proposed options would be great and thought about it some time ago, i will definitely take this and your scenarios into the todo list.

Thanks as well for the link to the documentation space, this looks great and exactly what i had in mind, i will take a look and see how this integrates with github.

**Reply by sidney-eliot** [reply]
Just finished the workflow part that could go into the documentation, it was quite fitting because I was anyway planning on writing it for my own wiki covering the character creation process. The page is [here](https://github.com/Epicrex/3DArtistsHandbook/wiki/ZBrush-To-Blender-Workflow). 

When you are at the point, where the documentation is ready. I can give you the markdown code of this wiki page, or convert it to reST if you'd rather use that

If you yourself aren't sure yet which of the 2 is better, reST or markdown. Markdown is more universal and easy to use but lacks helpful features like automatic tables of content for example.

**Reply by kromar** [reply]
this is great! i will let you know when i get to the documentation.

**Comment by sidney-eliot** [comment]
Hey @kromar, long time no see. Was curious if there were any developments on overhauling parts of the auto fix system, that we talked about back in mid 2022.
